### PR TITLE
Refactor abstract base test classes for integration tests

### PR DESCRIPTION
**Pull request by simonrozsival**

### CODE REVIEW

File: Toggl.Ultrawave.Tests.Integration/WorkspacesClientTests.cs [hunk 1]
```diff
@@ -30,12 +31,21 @@ protected override IObservable<List<Workspace>> CallEndpointWith(ITogglClient to
...
-            protected override int GetDefaultParameter(Ultrawave.User user, ITogglClient togglClient)
-                => user.DefaultWorkspaceId;
+            protected override IObservable<Workspace> CallEndpointWith(ITogglClient togglClient)
+            {
+                return Observable.Defer(async () =>
+                {
+                    var user = await togglClient.User.Get();
+                    return togglClient.Workspaces.GetById(user.DefaultWorkspaceId);
+                });
+            }
```
Comment (amulware) [inline Toggl.Ultrawave.Tests.Integration/WorkspacesClientTests.cs]: Please check this out especially, @willsb, since you know RX better than me.

File: Toggl.Ultrawave.Tests.Integration/WorkspacesClientTests.cs [hunk 2]
```diff
@@ -30,12 +31,21 @@ protected override IObservable<List<Workspace>> CallEndpointWith(ITogglClient to
...
+        public class TheGetByIdMethod : AuthenticatedGetEndpointBaseTests<Workspace>
         {
-            protected override int GetDefaultParameter(Ultrawave.User user, ITogglClient togglClient)
-                => user.DefaultWorkspaceId;
+            protected override IObservable<Workspace> CallEndpointWith(ITogglClient togglClient)
+            {
+                return Observable.Defer(async () =>
+                {
+                    var user = await togglClient.User.Get();
+                    return togglClient.Workspaces.GetById(user.DefaultWorkspaceId);
```
Comment (amulware) [inline Toggl.Ultrawave.Tests.Integration/WorkspacesClientTests.cs]: We are calling this in two places now, might be nicer to only call it in one (probably the one below?)

File: Toggl.Ultrawave.Tests.Integration/WorkspacesClientTests.cs [hunk 3]
```diff
@@ -30,12 +31,21 @@ protected override IObservable<List<Workspace>> CallEndpointWith(ITogglClient to
...
+                });
+            }
+
+            protected Func<Task> CallingEndpointWith(ITogglClient togglClient, int id)
```
Comment (amulware) [inline Toggl.Ultrawave.Tests.Integration/WorkspacesClientTests.cs]: This and the one below can be private.

File: Toggl.Ultrawave.Tests.Integration/Toggl.Ultrawave.Tests.Integration.csproj [hunk 4]
```diff
@@ -72,9 +72,10 @@
     <Reference Include="System.Xml.Linq" />
   </ItemGroup>
   <ItemGroup>
-    <Compile Include="EndpointTestBase.cs" />
-    <Compile Include="AuthenticatedEndpointBaseTests.cs" />
-    <Compile Include="AuthenticatedEndpointWithParameterBaseTests.cs" />
+    <Compile Include="BaseTests\AuthenticatedPostEndpointBaseTests.cs" />
+    <Compile Include="BaseTests\AuthenticatedGetEndpointBaseTests.cs" />
+    <Compile Include="BaseTests\EndpointTestBase.cs" />
+    <Compile Include="BaseTests\AuthenticatedEndpointBaseTests.cs" />
```
Comment (amulware) [inline Toggl.Ultrawave.Tests.Integration/Toggl.Ultrawave.Tests.Integration.csproj]: It would have been a bit easier to review if the files were renamed/moved in a different PR. ;)

File: Toggl.Ultrawave.Tests.Integration/WorkspacesClientTests.cs [hunk 5]
```diff
@@ -30,12 +31,21 @@ protected override IObservable<List<Workspace>> CallEndpointWith(ITogglClient to
...
         }
 
-        public class TheGetByIdMethod : AuthenticatedEndpointWithParameterBaseTests<Workspace, int>
+        public class TheGetByIdMethod : AuthenticatedGetEndpointBaseTests<Workspace>
         {
-            protected override int GetDefaultParameter(Ultrawave.User user, ITogglClient togglClient)
-                => user.DefaultWorkspaceId;
+            protected override IObservable<Workspace> CallEndpointWith(ITogglClient togglClient)
+            {
+                return Observable.Defer(async () =>
```
Comment (amulware) [inline Toggl.Ultrawave.Tests.Integration/WorkspacesClientTests.cs]: Consider using expression body syntax instead.

File: Toggl.Ultrawave.Tests.Integration/Toggl.Ultrawave.Tests.Integration.csproj (hunk 4 above)
Comment (simonrozsival) [inline Toggl.Ultrawave.Tests.Integration/Toggl.Ultrawave.Tests.Integration.csproj]: OK, I'll make a separate request next time

File: Toggl.Ultrawave.Tests.Integration/WorkspacesClientTests.cs [hunk 6]
```diff
@@ -30,12 +31,19 @@ protected override IObservable<List<Workspace>> CallEndpointWith(ITogglClient to
...
         }
 
-        public class TheGetByIdMethod : AuthenticatedEndpointWithParameterBaseTests<Workspace, int>
+        public class TheGetByIdMethod : AuthenticatedGetEndpointBaseTests<Workspace>
         {
-            protected override int GetDefaultParameter(Ultrawave.User user, ITogglClient togglClient)
-                => user.DefaultWorkspaceId;
+            protected override IObservable<Workspace> CallEndpointWith(ITogglClient togglClient)
+                => Observable.Defer(async () =>
+                    {
```
Comment (simonrozsival) [inline Toggl.Ultrawave.Tests.Integration/WorkspacesClientTests.cs]: I was not really sure what the indentation should be in this case, but it seams OK to me this way.

Comment (amulware) [inline Toggl.Ultrawave.Tests.Integration/WorkspacesClientTests.cs]: I feel the block should be indented one less, but that would not be entirely consistent to what we do in other places, so maybe this is fine.

**Comment by amulware** [comment]
@willsb Can you check if the line endings are fine now? It still shows a lot of changes that aren't changes.

**Comment by simonrozsival** [comment]
The only file which was affected by the CRLF problem in this PR was `StatusClientTests.cs`

**Comment by amulware** [comment]
What about `AuthenticatedEndpointBaseTests.cs` and `EndpointTestBase.cs`

All these three files still appear as completely rewritten in the diff.

**Comment by simonrozsival** [comment]
Yes they do seem like that - but that is probably because I moved them to a different directory. They are unchanged except a change in the namespace of `EndpointTestBase.cs`

File: Toggl.Ultrawave.Tests.Integration/WorkspacesClientTests.cs (hunk 6 above)
Comment (willsb) [inline Toggl.Ultrawave.Tests.Integration/WorkspacesClientTests.cs]: Change the one less indentation and this should be merge ready :)

Comment (simonrozsival) [inline Toggl.Ultrawave.Tests.Integration/WorkspacesClientTests.cs]: Fixed
